File: absNEON/R/calcAbsRatio.R
################################################################################
#' @title calcAbsRatio

#' @author
#' Robert Hensley \email{hensley@battelleecology.org} \cr

#' @description This function calculates the absorbance ratio for two user-
#' specified wavelengths from National Ecological Observatory Network (NEON)
#' water chemistry data.

#' @param absorbanceData User input of the table of NEON absorbance data [data.frame]
#' @param concentrationData User input of the table of NEON concentration data [data.frame]
#' @param wavelength1 User input of the wavelength to be used in the numerator [numeric]
#' @param wavelength2 User input of the wavelength to be used in the denominator [numeric]
#' @param correctFe User input of whether correction for overlapping absorbption of Fe(III)
#' should also be included. See README for more information. Defaults to FALSE. [logical]

#' @import tidyverse

#' @return This function returns a table of absorbance ratio values

#' @references
#' License: GNU AFFERO GENERAL PUBLIC LICENSE Version 3, 19 November 2007

#' @keywords NEON, DOC, SUVA, UV Absorbance

#' @examples
#' #Using an example file
#' #outputData <- calcAbsRatio(
#' #absorbanceData=swc_externalLabAbsorbanceScan,
#' #concentrationData=swc_externalLabDataByAnalyte,
#' #wavelength1 = 300
#' #wavelength2 = 400
#' #correctFe=TRUE)

#' @export

# changelog and author contributions / copyrights
################################################################################
calcAbsRatio<-function(
    absorbanceData,
    concentrationData=FALSE,
    wavelength1,
    wavelength2,
    correctFe=FALSE){

  if(nrow(absorbanceData) < 1){
    print("Error: No absorbance data")
    return(NULL)
  }

  if(wavelength1<220|wavelength1>600){
    print("Error: Wavelength 1 is outside 200-600nm range of absorbance scan")
    return(NULL)
  }
  if(wavelength2<220|wavelength2>600){
    print("Error: Wavelength 2 is outside 200-600nm range of absorbance scan")
    return(NULL)
  }

  # Isolates specified wavelengths if even, or surrounding wavelengths if odd.
  if(wavelength1 %% 2 == 0){
    absorbanceData1<-absorbanceData[(absorbanceData$wavelength==wavelength1),]
  }
  if(wavelength1 %% 2 != 0){
    absorbanceData1<-absorbanceData[(absorbanceData$wavelength==wavelength1-1|absorbanceData$wavelength==wavelength1+1),]
  }
  if(wavelength2 %% 2 == 0){
    absorbanceData2<-absorbanceData[(absorbanceData$wavelength==wavelength2),]
  }
  if(wavelength2 %% 2 != 0){
    absorbanceData2<-absorbanceData[(absorbanceData$wavelength==wavelength2-1|absorbanceData$wavelength==wavelength2+1),]
  }
  absorbanceData<-rbind(absorbanceData1,absorbanceData2)

  # Averages replicate absorbance scans
  absorbanceAveraged<-absorbanceData
  absorbanceAveraged$sampleID.wavelength<-paste(absorbanceAveraged$sampleID, absorbanceAveraged$wavelength, sep=".")
  absorbanceAveraged<-absorbanceAveraged |> dplyr::group_by(sampleID.wavelength) |> dplyr::summarize(
    sampleID=unique(sampleID),domainID=unique(domainID),siteID=unique(siteID),
    collectDate=unique(collectDate),wavelength=unique(wavelength),absorbance=mean(decadicAbsorbance))

  absorbanceAveraged$sampleID.wavelength<-NULL

  # No concentration required if Fe correction not performed
  combinedData<-absorbanceAveraged

  # Adds Fe concentrations if input condition is true
  if(correctFe == TRUE){
    Fe<-concentrationData[(concentrationData$analyte=="Fe"),]
    if(nrow(Fe) < 1){
      print("Error: No concentration data")
      return(NULL)
    }
    Fe<-Fe[,c("sampleID","analyteConcentration")]
    colnames(Fe)<-c("sampleID","Fe")
    combinedData<-merge(absorbanceAveraged, Fe,by.x="sampleID",by.y="sampleID")
    combinedData$absorbanceFe<-combinedData$Fe*((-0.00000044*(combinedData$wavelength^2))-(0.00007755*combinedData$wavelength)+0.11337671)
    combinedData$absorbanceCorrected<-combinedData$absorbance-combinedData$absorbanceFe
  }


  # Performs calculation without any corrections
    abs1<-combinedData[(combinedData$wavelength==wavelength1),]
    abs1<-abs1[,c("sampleID","absorbance")]
    names(abs1)[names(abs1) == "absorbance"] <- "abs1"
    abs2<-combinedData[(combinedData$wavelength==wavelength2),]
    abs2<-abs2[,c("sampleID","absorbance")]
    names(abs2)[names(abs2) == "absorbance"] <- "abs2"
    absRatio<-merge(abs1,abs2,by.x="sampleID",by.y="sampleID",all.x=T,all.y=T)
    absRatio$absRatio<-absRatio$abs1/absRatio$abs2


  # Performs calculation with Fe correction
  if(correctFe == TRUE){
    abs1Corrected<-combinedData[(combinedData$wavelength==wavelength1),]
    abs1Corrected<-abs1Corrected[,c("sampleID","absorbanceCorrected")]
    names(abs1Corrected)[names(abs1Corrected) == "absorbanceCorrected"] <- "abs1Corrected"
    abs2Corrected<-combinedData[(combinedData$wavelength==wavelength2),]
    abs2Corrected<-abs2Corrected[,c("sampleID","absorbanceCorrected")]
    names(abs2Corrected)[names(abs2Corrected) == "absorbanceCorrected"] <- "abs2Corrected"
    absRatioCorrected<-merge(abs1Corrected,abs2Corrected,by.x="sampleID",by.y="sampleID",all.x=T,all.y=T)
    absRatioCorrected$absRatioCorrected<-absRatioCorrected$abs1Corrected/absRatioCorrected$abs2Corrected
  }

  # Formats output table
  outputTable<-unique(absorbanceAveraged[,c("domainID","siteID","sampleID","collectDate")])
  outputTable<-merge(outputTable,absRatio,by.x="sampleID",by.y="sampleID",all.x=T,all.y=F)
  outputTable<-outputTable[,c("domainID","siteID","sampleID","collectDate","absRatio")]

  if(correctFe == TRUE){
    outputTable<-merge(outputTable,absRatioCorrected,by.x="sampleID",by.y="sampleID",all.x=T,all.y=F)
    outputTable<-outputTable[,c("domainID","siteID","sampleID","collectDate","absRatio","absRatioCorrected")]
  }

  return(outputTable)
}

File: neon-absorbance-py/src/neon_absorbance/calcAbsRatio.py
# Title: calcAbsRatio
#
# Original author: Robert Hensley <hensley@battelleecology.org>
# Translated to Python by: Adriana Teruel <teruel@battelleecology.org>
#
# Description: Calculate the absorbance ratio for two user-specified wavelengths
#   from NEON water chemistry scan data. Optionally applies Fe(III) correction.
#
# License: GNU AFFERO GENERAL PUBLIC LICENSE Version 3, 19 November 2007
#
# changelog and author contributions / copyrights
#   Adriana Teruel (2026-06-15)
#     original creation (Python translation of R package)

import pandas as pd

from ._helpers import _select_wavelength, _fe_absorbance


def calcAbsRatio(
    absorbanceData: pd.DataFrame,
    wavelength1: int,
    wavelength2: int,
    concentrationData: pd.DataFrame | None = None,
    correctFe: bool = False,
) -> pd.DataFrame:
    """
    Calculate the absorbance ratio for two user-specified wavelengths from
    National Ecological Observatory Network (NEON) water chemistry data.

    Parameters
    ----------
    absorbanceData : pd.DataFrame
        Table of NEON absorbance data (swc_externalLabAbsorbanceScan).
        Required columns: sampleID, domainID, siteID, collectDate,
        wavelength, decadicAbsorbance.
    wavelength1 : int
        Wavelength (nm) used as the numerator of the ratio. Must be in
        the 220–600 nm range. If odd, the two surrounding even wavelengths
        are averaged.
    wavelength2 : int
        Wavelength (nm) used as the denominator of the ratio. Must be in
        the 220–600 nm range. If odd, the two surrounding even wavelengths
        are averaged.
    concentrationData : pd.DataFrame, optional
        Table of NEON concentration data (swc_externalLabDataByAnalyte).
        Required when correctFe=True. Must contain columns: sampleID,
        analyte, analyteConcentration.
    correctFe : bool, optional
        Whether to apply a correction for the overlapping absorbance of
        Fe(III). Defaults to False. See README for details.

    Returns
    -------
    pd.DataFrame
        Table with columns: domainID, siteID, sampleID, collectDate,
        absRatio. If correctFe=True, also includes absRatioCorrected.

    Raises
    ------
    ValueError
        If absorbanceData is empty, either wavelength is outside 220–600 nm,
        or correctFe=True but no Fe concentration data is available.

    Examples
    --------
    >>> result = calcAbsRatio(
    ...     absorbanceData=swc_externalLabAbsorbanceScan,
    ...     wavelength1=254,
    ...     wavelength2=365,
    ... )

    >>> result = calcAbsRatio(
    ...     absorbanceData=swc_externalLabAbsorbanceScan,
    ...     wavelength1=254,
    ...     wavelength2=365,
    ...     concentrationData=swc_externalLabDataByAnalyte,
    ...     correctFe=True,
    ... )
    """
    if absorbanceData is None or absorbanceData.empty:
        raise ValueError("No absorbance data")

    if wavelength1 < 220 or wavelength1 > 600:
        raise ValueError("Wavelength 1 is outside 220-600 nm range")

    if wavelength2 < 220 or wavelength2 > 600:
        raise ValueError("Wavelength 2 is outside 220-600 nm range")

    absorbanceData1 = _select_wavelength(absorbanceData, wavelength1)
    absorbanceData2 = _select_wavelength(absorbanceData, wavelength2)

    combinedData = pd.concat(
        [absorbanceData1, absorbanceData2],
        ignore_index=True,
    )

    absorbanceAveraged = (
        combinedData
        .groupby(["sampleID", "wavelength"], as_index=False)
        .agg(
            domainID=("domainID", "first"),
            siteID=("siteID", "first"),
            collectDate=("collectDate", "first"),
            absorbance=("decadicAbsorbance", "mean"),
        )
    )

    combinedData = absorbanceAveraged.copy()

    if correctFe:
        if concentrationData is None or concentrationData.empty:
            raise ValueError("No concentration data")

        Fe = (
            concentrationData[concentrationData["analyte"] == "Fe"]
            [["sampleID", "analyteConcentration"]]
            .rename(columns={"analyteConcentration": "Fe"})
        )

        if Fe.empty:
            raise ValueError("No Fe concentration data")

        combinedData = combinedData.merge(Fe, on="sampleID", how="inner")
        combinedData["absorbanceFe"] = _fe_absorbance(
            combinedData["Fe"],
            combinedData["wavelength"],
        )
        combinedData["absorbanceCorrected"] = (
            combinedData["absorbance"] - combinedData["absorbanceFe"]
        )

    abs1 = (
        combinedData[combinedData["wavelength"] == wavelength1]
        [["sampleID", "absorbance"]]
        .rename(columns={"absorbance": "abs1"})
    )

    abs2 = (
        combinedData[combinedData["wavelength"] == wavelength2]
        [["sampleID", "absorbance"]]
        .rename(columns={"absorbance": "abs2"})
    )

    absRatio = abs1.merge(abs2, on="sampleID", how="outer")
    absRatio["absRatio"] = absRatio["abs1"] / absRatio["abs2"]

    outputTable = (
        absorbanceAveraged[["domainID", "siteID", "sampleID", "collectDate"]]
        .drop_duplicates()
        .merge(absRatio[["sampleID", "absRatio"]], on="sampleID", how="left")
    )

    if correctFe:
        abs1Corrected = (
            combinedData[combinedData["wavelength"] == wavelength1]
            [["sampleID", "absorbanceCorrected"]]
            .rename(columns={"absorbanceCorrected": "abs1Corrected"})
        )

        abs2Corrected = (
            combinedData[combinedData["wavelength"] == wavelength2]
            [["sampleID", "absorbanceCorrected"]]
            .rename(columns={"absorbanceCorrected": "abs2Corrected"})
        )

        absRatioCorrected = abs1Corrected.merge(
            abs2Corrected,
            on="sampleID",
            how="outer",
        )
        absRatioCorrected["absRatioCorrected"] = (
            absRatioCorrected["abs1Corrected"]
            / absRatioCorrected["abs2Corrected"]
        )

        outputTable = outputTable.merge(
            absRatioCorrected[["sampleID", "absRatioCorrected"]],
            on="sampleID",
            how="left",
        )

    return outputTable
